File: server/src/services/files.ts
import crypto from "node:crypto";
import fs from "node:fs/promises";
import path from "node:path";
import { v4 as uuidv4 } from "uuid";

import { config } from "../config.js";

export class FileServiceError extends Error {}

export async function ensureDataDirs() {
  await fs.mkdir(stegoDir(), { recursive: true });
}

export function stegoDir() {
  return path.join(config.dataDir, "stego");
}

export function safeFilename(filename: string | undefined, fallback = "file") {
  const base = path.basename(filename || fallback).trim().replace(/\s+/g, "_");
  const cleaned = base.replace(/[^A-Za-z0-9._-]+/g, "_").replace(/^[._]+|[._]+$/g, "");
  return (cleaned || fallback).slice(0, 120);
}

export function sha256Digest(data: Buffer) {
  return crypto.createHash("sha256").update(data).digest("hex");
}

export async function saveStegoFile(stegoBytes: Buffer, coverFilename: string) {
  await ensureDataDirs();
  const original = safeFilename(coverFilename, "cover");
  const stem = path.parse(original).name || "cover";
  const generatedName = `${stem}-${uuidv4().replaceAll("-", "").slice(0, 12)}.png`;
  const storagePath = path.join(stegoDir(), generatedName);
  await fs.writeFile(storagePath, stegoBytes);
  return { storagePath, generatedName };
}

export function requireUpload(file: Express.Multer.File | undefined, label: string) {
  if (!file?.buffer?.length) {
    throw new FileServiceError(`${label} was not uploaded.`);
  }
  if (file.size > config.maxUploadBytes) {
    throw new FileServiceError(`${label} is too large.`);
  }
  return file;
}

File: app/services/files.py
from __future__ import annotations

import hashlib
import uuid
from pathlib import Path

from fastapi import UploadFile

from app.config import Settings
from app.security import safe_filename


class FileServiceError(ValueError):
    pass


def ensure_data_dirs(settings: Settings) -> None:
    settings.data_dir.mkdir(parents=True, exist_ok=True)
    stego_dir(settings).mkdir(parents=True, exist_ok=True)


def stego_dir(settings: Settings) -> Path:
    return settings.data_dir / "stego"


def read_upload(upload: UploadFile, max_bytes: int) -> bytes:
    if not upload or not upload.filename:
        raise FileServiceError("No file was uploaded.")
    data = upload.file.read(max_bytes + 1)
    if len(data) > max_bytes:
        raise FileServiceError("Uploaded file is too large.")
    if not data:
        raise FileServiceError("Uploaded file is empty.")
    return data


def sha256_digest(data: bytes) -> str:
    return hashlib.sha256(data).hexdigest()


def save_stego_file(settings: Settings, stego_bytes: bytes, cover_filename: str) -> tuple[Path, str]:
    original = safe_filename(cover_filename, "cover")
    stem = Path(original).stem or "cover"
    generated_name = f"{stem}-{uuid.uuid4().hex[:12]}.png"
    path = stego_dir(settings) / generated_name
    path.write_bytes(stego_bytes)
    return path, generated_name

